### PR TITLE
feat: enhance error handling

### PR DESCRIPTION
**Pull request by PereiraE**

### CODE REVIEW

File: src/main/java/fr/uge/lp/databackbonepoc/controllers/error/ErrorResponse.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Objects;
+import lombok.Data;
+
+/**
+ * This class represent the generic response sent by the API in case of error.
+ */
+@Data
+public class ErrorResponse {
```
Comment (AxelBelin) [inline src/main/java/fr/uge/lp/databackbonepoc/controllers/error/ErrorResponse.java:12]: ErrorResponse might be a record.

File: src/main/java/fr/uge/lp/databackbonepoc/controllers/error/GlobalExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  @ExceptionHandler(AttributeNotFoundException.class)
+  @ResponseStatus(HttpStatus.NOT_FOUND)
+  public ResponseEntity<ErrorResponse> handleAttributeNotFoundException(AttributeNotFoundException ex) {
+    return notFound(ex.getMessage());
+  }
+
+  @ExceptionHandler(SamplerConfigurationNotFoundException.class)
+  @ResponseStatus(HttpStatus.NOT_FOUND)
```
Comment (AxelBelin) [inline src/main/java/fr/uge/lp/databackbonepoc/controllers/error/GlobalExceptionHandler.java:40]: ResponseStatus must be BAD_REQUEST

File: src/main/java/fr/uge/lp/databackbonepoc/controllers/error/GlobalExceptionHandler.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  public ResponseEntity<ErrorResponse> handleSamplerConfigurationNotFoundException(SamplerConfigurationNotFoundException ex) {
+    var response = new ErrorResponse(ex.getMessage());
+    return new ResponseEntity<>(response, HttpStatus.BAD_REQUEST);
+  }
+
+  @ExceptionHandler(InvalidResearchConfigurationException.class)
+  @ResponseStatus(HttpStatus.NOT_FOUND)
```
Comment (AxelBelin) [inline src/main/java/fr/uge/lp/databackbonepoc/controllers/error/GlobalExceptionHandler.java:47]: ResponseStatus must be CONFLICT

File: src/main/java/fr/uge/lp/databackbonepoc/controllers/error/GlobalExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+@RestControllerAdvice
+public class GlobalExceptionHandler {
+  private static ResponseEntity<ErrorResponse> notFound(String message) {
+    var response = new ErrorResponse(message);
+    return new ResponseEntity<>(response, HttpStatus.NOT_FOUND);
+  }
+
+  @ExceptionHandler(ResearchNotFoundException.class)
+  @ResponseStatus(HttpStatus.NOT_FOUND)
```
Comment (AxelBelin) [inline src/main/java/fr/uge/lp/databackbonepoc/controllers/error/GlobalExceptionHandler.java:22]: Is it necessary to specify a @ResponseStatus while you always return a NOT_FOUND status in notFound method ?

File: src/main/java/fr/uge/lp/databackbonepoc/linkresearch/sampler/Samplers.java [hunk 5]
```diff
@@ -14,16 +13,14 @@ public enum Samplers {
...
-    try {
-      return Optional.of(Samplers.valueOf(samplingAlgorithmName.toUpperCase(Locale.ROOT)));
-    } catch(IllegalArgumentException __) {
-      return Arrays.stream(Samplers.values())
-          .filter(enumKey -> enumKey.sampler.name().equalsIgnoreCase(samplingAlgorithmName))
-          .findAny();
-    }
+    var samplingAlgorithmNameToUpper = samplingAlgorithmName.toUpperCase(Locale.ROOT);
+    return Arrays.stream(Samplers.values())
+            .filter(samplers -> samplers.name().equals(samplingAlgorithmNameToUpper))
```
Comment (AxelBelin) [inline src/main/java/fr/uge/lp/databackbonepoc/linkresearch/sampler/Samplers.java:21]: .filter(enumKey -> enumKey.name().equals(samplingAlgorithmNameToUpper) || enumKey.sampler.name().equalsIgnoreCase(samplingAlgorithmNameToUpper))

**Review by AxelBelin — CHANGES_REQUESTED** [review]
Read comments

File: src/main/java/fr/uge/lp/databackbonepoc/controllers/error/ErrorResponse.java (hunk 1 above)
Comment (PereiraE) [reply src/main/java/fr/uge/lp/databackbonepoc/controllers/error/ErrorResponse.java:12]: The timestamp is not provided to the constructor.

File: src/main/java/fr/uge/lp/databackbonepoc/controllers/error/GlobalExceptionHandler.java (hunk 4 above)
Comment (PereiraE) [reply src/main/java/fr/uge/lp/databackbonepoc/controllers/error/GlobalExceptionHandler.java:22]: Yes it is, at least for documentation purpose with Swagger

File: src/main/java/fr/uge/lp/databackbonepoc/services/ResearchService.java [hunk 6]
```diff
@@ -227,22 +214,24 @@ public ResearchUpdateStatus addMatcherConfigurations(
...
     Objects.requireNonNull(configuredResearch);
+
     var leftSamplerConfig = configuredResearch.getLeftEntitySamplerConfiguration();
     if (leftSamplerConfig.getSamplingAlgorithm() == null) {
-      throw new IllegalArgumentException("Undefined sampling algorithm for left entity.");
+      throw new InvalidResearchConfigurationException("Missing sampling configuration for left entity.");
     }
 
     var rightSamplerConfig = configuredResearch.getRightEntitySamplerConfiguration();
     if(rightSamplerConfig.getSamplingAlgorithm() == null) {
```
Comment (julienleows) [inline src/main/java/fr/uge/lp/databackbonepoc/services/ResearchService.java:228]:
```suggestion
    if (rightSamplerConfig.getSamplingAlgorithm() == null) {
```

**Review by AxelBelin — APPROVED** [review]

**Pull request merged**